<!Doctype html>
<html>
    <head>
        <meta charset="UTF-8" />
        <title>BOSTIn</title>
        <link rel="stylesheet" href="style.css"/>
    </head>
    <body>
        <a href="https://github.com/JFFleming/BOSTIn/">
            <svg width="80" height="80" viewBox="0 0 250 250" style="fill:white; color:black; z-index: 51; position: absolute; top: 0; right: 0; border: 0;" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path>
            </svg>
        </a>
        <div class="logo">
            <center><img src="images/logo.png" width="400" height="400"></center>
         </div>

        <div class="contentdiv">
            <center><h1>BOSTIn</h1>
            <p id="cite">To cite this program and for more information, see <a href="https://github.com/JFFleming/BOSTIn/tree/main">Fleming & Struck 2024 Citation!</a></p>
            </center>
            <div class="navigation">
                <a href="faq.html">FAQ</a>
                <a href="mailto: j.f.fleming@nhm.uio.no">E-Mail the authors</a>
            </div>
            <h2></h2>
            <br>
            <strong> BOSTIn (Broad Overview of Sequence & Topology Incongruence) </strong> is a new, user-friendly phylogenetic artifact identification package.
            <br><br>
        	It takes an aligned fasta file and uses up to three different metrics to assess the alignment for compositional heterogeneity, site saturation and branch length heterogeneity.
        	<br>
        	In addition, it then produces a unique "narrative report", which tells you which taxa are potentially concerning, and why. It also presents possible next steps to ameliorate those artifacts, or explore them in more detail.
        	<br>
        	For Compositional Heterogeneity, BOSTIn uses nRCFV, which measures how variable nucleotide or amino acid composition is across the dataset. While all three measurements give taxon specific ratings as well as whole dataset ratings, nRCFV also gives a character specific measurement.
        	<br>
        	For Site Saturation, BOSTIn uses the DEScore for amino acid data and the C-Score for nucleotide data. The former measures the proportion of between Dayhoff category changes across a dataset, while the latter compares the transition/transversion ratio to the distance between genes in the dataset (the uncorrected p-distance)
            <br>
			For Branch Length Heterogeneity, BOSTIn uses the LB-Score, which compares the variation of distances between taxa across the tree.
            <br>
            To find out more about how all these metrics are calculated, you can visit our citations page <a href="https://github.com/JFFleming/BOSTIn/tree/main"> here </a>
            <br><br>
            To get started, simply upload an aligned fasta file below, note whether it is nucleotide or amino acid data, and then tell BOSTIn which artifacts you are interested in exploring today by clicking the text boxes. BOSTIn can run all three analyses at once, or any combination of the three separately!</td> <br><br>
            <span class="middle">
                <input id="faFile" type="file" autocomplete="off">
            </span>
        <br>
            <br>
            <label for="type">Alignment Datatype:</label>
			<select name="dtype" id="type">
			  <option value="none">Select Option</option>
			  <option value="nucl">Nucleotide</option>
			  <option value="prot">Amino Acid</option>
			</select>
			<br>
			<br>
			Phylogenetic Artifacts:
			<br>
            <input id="CompHet" type="checkbox" autocomplete="off"/>Compositional Heterogeneity (nRCFV)<br>
            <input id="SiteSat" type="checkbox" autocomplete="off"/>Site Saturation (DE-Score on amino acids, C-Score on nucleotides)<br>
            <input id="BLH" type="checkbox" autocomplete="off"/>Branch Length Heterogeneity (LB-Score)<br>
            <br><br>
            <span class="middle">
                <button onclick="bostin()">Run BOSTIn</button>
                <button onclick="resetForm()">Reset form</button>
            </span>
            <div class="navigation">
                <a href="faq.html">FAQ</a>
                <a href="mailto: j.f.fleming@nhm.uio.no">E-Mail the author</a>
            </div>
        </div>
<script src="js/results.js"></script>
<script src="js/blh.js"></script>
<script src="js/sat_aa.js"></script>
<script src="js/comphet.js"></script>

    </body>
</html>
